### PR TITLE
Fix problems with claimId in performStepVerification

### LINKED ISSUE #73: in performStepVerification, any claimID can be passed
Might mess things up in the claim manager.

### PR DESCRIPTION
**Pull request by mrsmkl**
The new way of generating the claimId isn't working, yet.
Going to revert some changes next.
Should fix #73

### CODE REVIEW

File: contracts/ClaimManager.sol [hunk 1]
```diff
@@ -171,19 +181,19 @@ contract ClaimManager is DepositsManager {
     require(claimExists(claim));
     require(!claim.decided);
 
+    require(claim.verificationOngoing == false);
+
     // check if there is a challenger who has not the played verification game yet.
     if (claim.numChallengers > claim.currentChallenger) {
-      require(claim.verificationOngoing == false);
```
Comment (Shrugs) [inline contracts/ClaimManager.sol]: are we sure we want to remove this line?

File: contracts/ClaimManager.sol [hunk 2]
```diff
@@ -121,9 +126,15 @@ contract ClaimManager is DepositsManager {
     }
 
     require(deposits[claimant] >= minDeposit);
-    require(claimantClaims[claimant] == 0);//claimant can only do one claim at a time
+    // require(claimantClaims[claimant] == 0);//claimant can only do one claim at a time
+    
+//    uint claimId = numClaims;
```
Comment (shrugs) [inline contracts/ClaimManager.sol]: we can remove these comments

File: contracts/ClaimManager.sol [hunk 3]
```diff
@@ -133,10 +144,10 @@ contract ClaimManager is DepositsManager {
     claim.createdAt = block.number;
     claim.decided = false;
     claim.proposalId = proposalId;
-    claimantClaims[claimant] = numClaims;
+    // claimantClaims[claimant] = claimId;
```
Comment (shrugs) [inline contracts/ClaimManager.sol]: nitpick: remove

File: contracts/ClaimManager.sol [hunk 4]
```diff
@@ -197,6 +207,8 @@ contract ClaimManager is DepositsManager {
   // @param sessionId – the sessionId.
   // @param winner – winner of the verification game.
   // @param loser – loser of the verification game.
+  
+  // where does the claim id come from?
```
Comment (shrugs) [inline contracts/ClaimManager.sol]: @sinahab this comes from the DogeRelay right?

File: contracts/ClaimManager.sol [hunk 5]
```diff
@@ -253,7 +265,7 @@ contract ClaimManager is DepositsManager {
     require(claim.decided);
 
     unbondDeposit(claimID, claim.claimant);
-    claimantClaims[claim.claimant] = 0;
+    // claimantClaims[claim.claimant] = 0;
```
Comment (shrugs) [inline contracts/ClaimManager.sol]: nitpick: remove

File: contracts/Verifier.sol [hunk 6]
```diff
@@ -21,6 +21,7 @@ contract Verifier {
 
     struct VerificationSession {
         uint id;
+        // uint claimid;
```
Comment (shrugs) [inline contracts/Verifier.sol]: nitpick: remove

File: contracts/Verifier.sol [hunk 7]
```diff
@@ -52,23 +55,23 @@ contract Verifier {
...
-            lastClaimantMessage: now,
-            lastChallengerMessage: now,
-            lowStep: 0,
-            lowHash: keccak256(_input),
-            medStep: 0,
-            medHash: 0,
-            highStep: steps,
-            highHash: keccak256(_output)
-        });
+        VerificationSession storage s = sessions[sessionId];
```
Comment (shrugs) [inline contracts/Verifier.sol:58]: is there a reason we changed the format of this?

File: contracts/Verifier.sol [hunk 8]
```diff
@@ -36,9 +37,11 @@ contract Verifier {
     }
 
     mapping(uint => VerificationSession) public sessions;
+    mapping(uint => uint) public sessions_claimid;
```
Comment (shrugs) [inline contracts/Verifier.sol]: let's make this `camelCase` for linting purposes

File: client/db/config/config.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
   "development": {
     "username": "postgres",
-    "password": null,
+    "password": "asd",
```
Comment (hswick) [inline client/db/config/config.json]: I think this will break on our machines?

File: contracts/ClaimManager.sol [hunk 10]
```diff
@@ -109,6 +109,10 @@ contract ClaimManager is DepositsManager {
   // @param _plaintext – the plaintext blockHeader.
   // @param _blockHash – the blockHash.
   // @param claimant – the address of the Dogecoin block submitter.
+  function calcId(bytes _plaintext, bytes32 _hash, address claimant, bytes32 proposalId) public pure returns (uint) {
```
Comment (hswick) [inline contracts/ClaimManager.sol]: I think this could be moved to scryptRunner where it can be computed offchain

Comment (shrugs) [reply contracts/ClaimManager.sol]: do we need to trust that it was computed correctly?

File: contracts/ClaimManager.sol [hunk 11]
```diff
@@ -121,9 +126,15 @@ contract ClaimManager is DepositsManager {
     }
 
     require(deposits[claimant] >= minDeposit);
-    require(claimantClaims[claimant] == 0);//claimant can only do one claim at a time
+    // require(claimantClaims[claimant] == 0);//claimant can only do one claim at a time
+    
+//    uint claimId = numClaims;
+//    uint claimId = uint(keccak256(claimant, _plaintext, _hash, numClaims));
+
+    uint claimId = uint(keccak256(claimant, _plaintext, _hash, proposalId));
```
Comment (hswick) [inline contracts/ClaimManager.sol]: I think we could get away with not hashing the plaintext. that way total parameter size is 96 bytes. As opposed to plaintext being dynamically sized

File: contracts/ClaimManager.sol (hunk 10 above)
Comment (hswick) [reply contracts/ClaimManager.sol]: Well users can pass in the id's anyways. Doesnt matter if its done in scryptVerifier.calcId or in scryptRunner. Better yet we could use web3.utils.sha3

File: contracts/Verifier.sol (hunk 7 above)
Comment (mrsmkl) [reply contracts/Verifier.sol:58]: I tried to work around stack limitation, but it didn't work out.

File: contracts/ClaimManager.sol (hunk 11 above)
Comment (mrsmkl) [reply contracts/ClaimManager.sol]: Good idea.

File: contracts/ClaimManager.sol [hunk 12]
```diff
@@ -109,8 +109,8 @@ contract ClaimManager is DepositsManager {
   // @param _plaintext – the plaintext blockHeader.
   // @param _blockHash – the blockHash.
   // @param claimant – the address of the Dogecoin block submitter.
-  function calcId(bytes _plaintext, bytes32 _hash, address claimant, bytes32 proposalId) public pure returns (uint) {
-    return uint(keccak256(claimant, _plaintext, _hash, proposalId));
+  function calcId(bytes, bytes32 _hash, address claimant, bytes32 proposalId) public pure returns (uint) {
```
Comment (hswick) [inline contracts/ClaimManager.sol]: why is there a bytes at beginning of method?

**Comment by sinahab** [comment]
Nice. This looks great!

Either before or after the merge, we should write a test that demonstrates parallel claims being processed. Don't think we need to hold up the merge on this though, we can do it after.

File: contracts/Verifier.sol (hunk 7 above)
Comment (shrugs) [reply contracts/Verifier.sol:58]: makes sense, thanks

**Review by shrugs — APPROVED** [review]
lgtm, agreed on tests

**Pull request merged**